j5/boards/arduino/uno.py

```diff
@@ -0,0 +1,114 @@
+"""Classes for the Arduino Uno."""
+from enum import IntEnum
+from typing import Mapping, Optional, Set, Type, Union, cast
+
+from j5.backends import Backend
+from j5.boards import Board
+from j5.components import (
+    LED,
+    Component,
+    GPIOPin,
+    GPIOPinInterface,
+    GPIOPinMode,
+    LEDInterface,
+)
+
+
+class AnaloguePin(IntEnum):
+    """Analogue Pins numbering."""
+
+    A0 = 14
+    A1 = 15
+    A2 = 16
+    A3 = 17
+    A4 = 18
+    A5 = 19
+
+
+PinNumber = Union[int, AnaloguePin]
```

Comment:
Reckon we should make a `DigitalPin` enum too, since there's a finite number of them and pins `0` and `1` are invalid? Or do you think that's more trouble than it's worth? I don't have a strong opinion on this.

Reply:
> pins 0 and 1 are invalid

I'd argue that's a implementation details that should throw an exception in the backend. e.g a different backend / comms protocol could be using i2c and thus block `A4` and `A5`.

> should make a DigitalPin enum too

In principle, I like this suggestion. My one concern is what we would call the entries?

e.g

We can't use `4` as that is a protected thingy in Python.

Maybe `D2` - `D13`? Although that then doesn't match the numbering on the board.

Reply:
> I'd argue that's a implementation details that should throw an exception in the backend. e.g a different backend / comms protocol could be using i2c and thus block `A4` and `A5`.

Didn't think of this, yeah I agree.

> Maybe `D2` - `D13`?

This is what I was thinking too, but yeah the labelling mismatch might cause confusion.

Maybe we don't need to go the full mile here - we could just use a [type alias](https://mypy.readthedocs.io/en/latest/kinds_of_types.html#type-aliases) for `int` rather than defining a new enum.

Reply:
> Maybe we don't need to go the full mile here - we could just use a [type alias](https://mypy.readthedocs.io/en/latest/kinds_of_types.html#type-aliases) for `int` rather than defining a new enum.

What's your opinion on using a [NewType](https://mypy.readthedocs.io/en/latest/more_types.html)?

Reply:
> What's your opinion on using a [NewType](https://mypy.readthedocs.io/en/latest/more_types.html)?

That would be even better. (It was also my original intention, but I wasn't sure whether they existed in mypy and failed to find it in the docs)

Reply:
I've attempted this and it would require students to do `pins[DigitalPin(4)]`, which is kinda nasty UX, so let's not.